File: 0212-Word Search II/0212-Word Search II.cpp
class TrieNode {
public:
	TrieNode() {
		memset(next, 0, sizeof(next));
		word = "";
	}
    TrieNode *next[26];
    string word;
};

class Solution {
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        TrieNode *root = buildTrie(words);
        vector<string> result;
        for (int i = 0; i < board.size(); ++i)
        {
            for (int j = 0; j < board[0].size(); ++j)
            {
                dfs(board, i, j, root, result);
            }
        }
        
        return result;
    }
    
    void dfs(vector<vector<char>>& board, int i, int j, TrieNode *root, vector<string>& result)
    {
        char c = board[i][j];
        if (c == '#' || root->next[c - 'a'] == nullptr)
        {
            return;
        }
        
        root = root->next[c - 'a'];
        if(root->word != "")
        {
            result.push_back(root->word);
            root->word = "";
        }
        board[i][j] = '#';
        if (i > 0)
        {
            dfs(board, i - 1, j, root, result);
        }
        if (i < board.size() - 1)
        {
            dfs(board, i + 1, j, root, result);
        }
        if (j > 0)
        {
            dfs(board, i, j - 1, root, result);
        }
        if (j < board[0].size() - 1)
        {
            dfs(board, i, j + 1, root, result);
        }
        board[i][j] = c;
    }
    
    TrieNode* buildTrie(vector<string>& words)
    {
        TrieNode *root = new TrieNode();
        for (string w : words)
        {
            TrieNode *p = root;
            for (char c : w)
            {
                int i = c - 'a';
                if (p->next[i] == nullptr)
                {
                    p->next[i] = new TrieNode();
                }
                p = p->next[i];
            }
            
            p->word = w;
        }
        
        return root;
    }
};

File: 0201-0300/0212-Word Search II/0212-Word Search II.py
import collections
class TrieNode:
    
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.word = ''

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        root = TrieNode()
        for word in words:
            node = root
            for c in word:
                node = node.children[c]
            node.word = word
        
        result = []
        def dfs(i, j, node):
            c = board[i][j]
            if c == '#' or c not in node.children:
                return
            node = node.children[c]
            if node.word != '':
                result.append(node.word)
                node.word = ''
                
            board[i][j] = '#'
            if i > 0:
                dfs(i - 1, j, node)
            if i < len(board) - 1:
                dfs(i + 1, j, node)
            if j > 0:
                dfs(i, j - 1, node)
            if j < len(board[0]) - 1:
                dfs(i, j + 1, node)
            board[i][j] = c
            
        for i in range(len(board)):
            for j in range(len(board[0])):
                dfs(i, j, root)

        return result
